### PR TITLE
MAISTRA-2051: Allow istiod to run without Node access

### PR DESCRIPTION
**Pull request by bison**
This is a reimplementation of the following for the Maistra 2.1 /
Istio 1.8 rebase:

 - MAISTRA-1723: Allow disabling NodePort gateway support (#180)
 - Ensure Pilot can run without privileges for reading nodes (#125)

A new istiod flag is added, `disableNodeAccess`, that will disable any
feature relying on access to Node objects.

### CODE REVIEW

**Comment by bison** [comment]
cc: @rcernich @luksa

File: pilot/cmd/pilot-discovery/main.go [hunk 1]
```diff
@@ -149,6 +149,8 @@ func init() {
 		"The name of the MemberRoll resource")
 	discoveryCmd.PersistentFlags().BoolVar(&serverArgs.RegistryOptions.KubeOptions.EnableCRDScan, "enableCRDScan", true,
 		"Whether to scan CRDs at startup")
+	discoveryCmd.PersistentFlags().BoolVar(&serverArgs.RegistryOptions.KubeOptions.DisableNodeAccess, "disableNodeAccess", false,
+		"Whether to prevent istiod watching Node objects")
```
Comment (rcernich) [inline pilot/cmd/pilot-discovery/main.go:153]: Previously, we've been using `podLocalitySource=pod` for this.  Is there a reason to change it?

Comment (bison) [reply pilot/cmd/pilot-discovery/main.go:153]: Well, this has all changed since the 1.6 release. In that particular case, we don't really need the pod locality thing anymore, as Istio already has an annotation that goes on Pods that indicates locality. So this checks that, and if it doesn't exist, it falls back looking for the Node labels that have been copied to the Pod -- if you've disabled Node access. Otherwise it goes on to look at the Nodes. Disabling the Node informer that would normally be used to check the Node labels for locality there also has the effect of disabling support for NodePort gateway because we can't read the addresses. And potentially other features in the future.

Basically if you disable the Node informer in the controller, you're affecting multiple features that might rely on it. So it seems better to just have a single flag for it.

Comment (rcernich) [reply pilot/cmd/pilot-discovery/main.go:153]: Okay.  We'll need to make changes to the istiod charts to accommodate the new flag.

Comment (bison) [reply pilot/cmd/pilot-discovery/main.go:153]: We may also want to have the operator start preferring the upstream locality label on pods:

https://github.com/istio/istio/blob/1.8.0/pilot/pkg/model/service.go#L139-L146

Comment (rcernich) [reply pilot/cmd/pilot-discovery/main.go:153]: I don't think we're modifying any charts related to those.

**Review by rcernich — APPROVED** [review]

File: pilot/cmd/pilot-discovery/main.go (hunk 1 above)
Comment (bison) [reply pilot/cmd/pilot-discovery/main.go:153]: In case it wasn't clear, I mean right now our operator copies the topology information from Node objects to Pods, right? Istio now has a dedicated label for Pods to read that information. So if we change the operator to start writing that new label to pods, we can eventually drop almost all of this and be more inline with upstream.

Comment (rcernich) [reply pilot/cmd/pilot-discovery/main.go:153]: Ah, understood.  Could you create an issue for 2.1 please?

Comment (bison) [reply pilot/cmd/pilot-discovery/main.go:153]: https://issues.redhat.com/browse/MAISTRA-2115

Comment (rcernich) [reply pilot/cmd/pilot-discovery/main.go:153]: Thanks @bison!
